server/server.go

```diff
@@ -575,20 +580,20 @@ func (c *conn) SubmitResult(result interface{}) (ok bool) {
 	if err != nil {
 		c.LogConnErr(err)
 		c.conn.Close()
+		atomic.StoreUint32(&c.closed, 1)
 		return false
 	}
 	return true
 }
 
 func (c *conn) IsAlive() bool {
-	var buf [0]byte
-	_, err := c.conn.Read(buf[:])
-	return err == nil
+	return atomic.LoadUint32(&c.closed) == 0
 }
```

Comment:
Checking `c.closed` shouldn't replace the existing check, but rather should be performed in addition - `IsAlive` should return false if either `err != nil` OR if `c.closed` is 0. The reason for this is that the connection can become closed without being explicitly closed by our code - if the other side closes the connection.

Reply:
Reading with an empty buffer will never return an error.

Reply:
Oh, now I see what you mean about `tls.Conn`'s `Read` method. OK.

Reply:
But I agree with @lziest - we should probably include it anyway in case the behavior changes in the future. Can't hurt.

Reply:
If we are tracking errors manually, I don't believe there is any point in doing both.